daml/ContingentClaims/Lifecycle.daml

```diff
@@ -42,56 +41,45 @@ data Result a = Result with
 -- 3. A set of `Claim`s.
 -- 4. A function which, given a date, lifecycles the third argument.
 lifecycle : (Eq a, CanAbort m)
-  => (Text -> Date -> m Decimal) 
-  -> (C a -> C a -> C a) 
-  -> C a 
-  -> Date -> m (Result a) 
+  => (Text -> Date -> m Decimal)
+  -> (C a -> C a -> C a)
+  -> C a
+  -> Date -> m (Result a)
 lifecycle spot choose
   = runKleisli
   . fmap (uncurry Result)
-  . flip runReaderT 1.0
+  . flip evalStateT 1.0
   . runWriterT
-  . para (lifecycle' spot choose)
-
-type Seed m a = WriterT [(Decimal, a)] (ReaderT Decimal (Kleisli m Date)) (C a)
-
--- TODO: these three should probably go in their respective modules
+  . apoCataM (pruneZeros') (lifecycle' spot choose)
 
-instance (Monoid w, CanAbort m) => CanAbort (WriterT w m) where
-  abort = lift . abort
-
-instance CanAbort m => CanAbort (ReaderT s m) where
-  abort = lift . abort
-
-instance CanAbort m => CanAbort (Kleisli m a) where -- because of the order of parameters it's not possible to write a MonadTrans instance for Kleisli.
-  abort = Kleisli . const . abort
+type Seed m a = WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (Either (C a) (C a)))
 
 lifecycle' : (Eq a, Monad m, CanAbort m)
   => (Text -> Date -> m Decimal)
   -> (C a -> C a -> C a)
-  -> F a (C a, Seed m a) -> Seed m a
+  -> C a -> Seed m a
 
-lifecycle' _ _ ZeroF = pure Zero
-lifecycle' _ _ (OneF id) = do
-  qty <- ask
+lifecycle' _ _ Zero = pure ZeroF
+lifecycle' _ _ (One id) = do
+  qty <- get
   tell [(qty, id)]
-  pure $ One id
-lifecycle' _ _ (GiveF (_, m)) = local negate m
-lifecycle' spot _ (WhenF obs (c, m)) = do
-  wasLifecycled <- lift . lift $ eval spot obs
-  if wasLifecycled then m $> Zero else pure $ When obs c
-lifecycle' spot _ (ScaleF obs (_, m)) = do
+  pure ZeroF -- replace lifecycled node with zero, to avoid double-spend.
+lifecycle' _ _ (Give c) = do
+  modify negate 
```

Comment:
I changed the `ReaderT` to a `StateT`. When I re-wrote the recursion to be top-down, I started having problems with the reader. I'm not certain if this was somehow due to the semantics (the `local` operation from `ReaderT` is actually contra-map - I don't know if this somehow breaks if you're using corecursion instead of recursion). But overall, it's more ledgible, so I decided to stick with this.
The only doubt I have is whether this may be less efficient (maybe the function in reader is treated as a first-class lambda in the JVM, instead of the tuple `(,)`, which probably requires an instantiation - but just guessing here). Don't think it matters in the big scheme of things, but thought I'd document it here.